core/src/main/java/org/apache/carbondata/core/carbon/datastore/BlockIndexStore.java

```diff
@@ -16,113 +16,142 @@
-    blockInfoLock = new ConcurrentHashMap<BlockInfo, Object>();
-    segmentIdToBlockListMap = new ConcurrentHashMap<>();
-    mapOfBlockInfoToFuture = new ConcurrentHashMap<>();
+    // if block is not loaded
+    if (null == tableBlock) {
+      // check any lock object is present in
+      // block info lock map
+      Object blockInfoLockObject = blockInfoLock.get(blockInfo);
+      // if lock object is not present then acquire
+      // the lock in block info lock and add a lock object in the map for
+      // particular block info, added double checking mechanism to add the lock
+      // object so in case of concurrent query we for same block info only one lock
+      // object will be added
+      if (null == blockInfoLockObject) {
+        synchronized (blockInfoLock) {
+          // again checking the block info lock, to check whether lock object is present
+          // or not if now also not present then add a lock object
+          blockInfoLockObject = blockInfoLock.get(blockInfo);
+          if (null == blockInfoLockObject) {
+            blockInfoLockObject = new Object();
+            blockInfoLock.put(blockInfo, blockInfoLockObject);
+          }
+        }
+      }
+      //acquire the lock for particular block info
+      synchronized (blockInfoLockObject) {
+        // check again whether block is present or not to avoid the
+        // same block is loaded
+        //more than once in case of concurrent query
+        tableBlock = (AbstractIndex) lruCache.get(
+            getLruCacheKey(tableBlockUniqueIdentifier.getAbsoluteTableIdentifier(), blockInfo));
+        // if still block is not present then load the block
+        if (null == tableBlock) {
+          tableBlock = loadBlock(tableBlockUniqueIdentifier);
+          fillSegmentIdToBlockListMap(tableBlockUniqueIdentifier.getAbsoluteTableIdentifier(),
+              blockInfo);
+        }
+      }
+    } else {
+      tableBlock.incrementAccessCount();
+    }
+    return tableBlock;
   }
 
   /**
-   * Return the instance of this class
-   *
-   * @return singleton instance
+   * @param absoluteTableIdentifier
+   * @param blockInfo
    */
-  public static BlockIndexStore getInstance() {
-    return CARBONTABLEBLOCKSINSTANCE;
+  private void fillSegmentIdToBlockListMap(AbsoluteTableIdentifier absoluteTableIdentifier,
+      BlockInfo blockInfo) {
+    TableSegmentUniqueIdentifier segmentIdentifier =
+        new TableSegmentUniqueIdentifier(absoluteTableIdentifier,
+            blockInfo.getTableBlockInfo().getSegmentId());
+    List<BlockInfo> blockInfos =
+        segmentIdToBlockListMap.get(segmentIdentifier.getUniqueTableSegmentIdentifier());
+    if (null == blockInfos) {
+      synchronized (segmentIdentifier) {
+        blockInfos =
+            segmentIdToBlockListMap.get(segmentIdentifier.getUniqueTableSegmentIdentifier());
+        if (null == blockInfos) {
+          blockInfos = new CopyOnWriteArrayList<>();
```

Comment:
Why we need copyOnWriteArrayList?? As this list contain all the block info for one segment. So when we are taking a lock for segment i think copyonwrite array list is not required. Please comment